### PR TITLE
[Support] Add class for generating unique names based on a symbol cache

### PR DESCRIPTION
**Pull request by mortbopet**
This commit introduces a new utility class for generating unique names based on some input base name. The logic for permutating the base name can be user-specified, but defaults to a simple inter-suffixed scheme.

As a sample use, HWSpecialize has been rewritten to use the SymbolUniquer.

### CODE REVIEW

**Comment by teqdruid** [comment]
Are there any cases wherein something other than an integer is used as the suffix? If not, I would just replace the `any` with `uint64_t` and make it more generic when the need arises.

**Comment by teqdruid** [comment]
Are there any cases wherein something other than an integer is used as the suffix? If not, I would just replace the `any` with `uint64_t` and make it more generic when the need arises.

File: include/circt/Support/SymUniquer.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
+//===- SymUniquer.h - Declare Symbol Uniquer --------------------*- C++ -*-===//
```
Comment (mikeurbach) [inline include/circt/Support/SymUniquer.h:1]: nit: but I'd spell out SymbolUniquer in the filename if that's what the main class is named. I guess I would have said the same about SymCache, but missed that.

**Review by mikeurbach — COMMENTED** [review]
A couple thoughts here:

1. I am also wondering the the value of parameterizing by the uniquerFn, why not just start with the integer increment approach and generalize if anyone ever wants something different.
2. I am wondering about re-use here. We already have a Namespace utility in Support, which achieves something similar: https://github.com/llvm/circt/blob/90ae0886aa06020d9f370118bb69dc1d3afbddcb/include/circt/Support/Namespace.h. Could that be rewritten in terms of the symbol cache and uniquer? There are also [two](https://github.com/llvm/circt/blob/9286f2f83b196a1477fc4b57538b4e1a518350b6/include/circt/Dialect/FIRRTL/Namespace.h) [other](https://github.com/llvm/circt/blob/03458d7fa401cc9d62248ceb9e01a336abe21391/include/circt/Dialect/HW/Namespace.h) namespace helpers. It would be great to unify all of this and not just add a fourth thing.

**Comment by mortbopet** [comment]
> I am wondering about re-use here. We already have a Namespace utility in Support, which achieves something similar:

Fully agree; there is no argument for adding this utility with the other ones in tree. I'll do a bit of thinking to figure out how to best unify the behaviours of the two (four) implementations.

**Comment by mortbopet** [comment]
Closing this in favor of unifying support based on the existing `Namespace` class.

**Pull request closed without merging**